flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java

```diff
@@ -0,0 +1,109 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.changelog;
+
+import org.apache.flink.api.common.state.State;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.state.changelog.StateChange;
+import org.apache.flink.runtime.state.internal.InternalKvState;
+import org.apache.flink.runtime.state.internal.InternalValueState;
+
+import java.io.IOException;
+
+/**
+ * Delegated partitioned {@link ValueState} that forwards changes to {@link StateChange} upon {@link
+ * ValueState} is updated.
+ *
+ * @param <K> The type of the key.
+ * @param <N> The type of the namespace.
+ * @param <V> The type of the value.
+ */
+class ChangelogValueState<K, N, V> implements InternalValueState<K, N, V> {
+    private final InternalValueState<K, N, V> valueState;
+
+    ChangelogValueState(InternalValueState<K, N, V> valueState) {
+        this.valueState = valueState;
+    }
+
+    @Override
+    public V value() throws IOException {
+        return valueState.value();
+    }
+
+    @Override
+    public void update(V value) throws IOException {
+        valueState.update(value);
+    }
+
+    @Override
+    public TypeSerializer<K> getKeySerializer() {
+        return valueState.getKeySerializer();
+    }
+
+    @Override
+    public TypeSerializer<N> getNamespaceSerializer() {
+        return valueState.getNamespaceSerializer();
+    }
+
+    @Override
+    public TypeSerializer<V> getValueSerializer() {
+        return valueState.getValueSerializer();
+    }
+
+    @Override
+    public void setCurrentNamespace(N namespace) {
+        valueState.setCurrentNamespace(namespace);
+    }
+
+    @Override
+    public byte[] getSerializedValue(
+            byte[] serializedKeyAndNamespace,
+            TypeSerializer<K> safeKeySerializer,
+            TypeSerializer<N> safeNamespaceSerializer,
+            TypeSerializer<V> safeValueSerializer)
+            throws Exception {
+        return valueState.getSerializedValue(
+                serializedKeyAndNamespace,
+                safeKeySerializer,
+                safeNamespaceSerializer,
+                safeValueSerializer);
+    }
+
+    @Override
+    public StateIncrementalVisitor<K, N, V> getStateIncrementalVisitor(
+            int recommendedMaxNumberOfReturnedRecords) {
+        return valueState.getStateIncrementalVisitor(recommendedMaxNumberOfReturnedRecords);
+    }
+
+    @Override
+    public void clear() {
+        valueState.clear();
+    }
+
+    public InternalValueState<K, N, V> getDelegatedValueState() {
```

Comment:
Actually, I think all such `getDelegatedValueState` could be simplified to `getDelegatedState`, and we always keep just protected `delegatedState` within the `AbstractChangelogState`.

Reply:
Same as above.